### PR TITLE
refactor: extract DML types (#2731)

### PR DESCRIPTION
**Pull request by tustvold**
As part of routing deletes through the write buffer, I need to define an in-memory type for the data that is being sequenced. Following internal discussion we decided to call this DML to be consistent with other database's name for this.

It didn't seem write to define these types in `mutable_batch` as it has no concept of a delete, but it also seemed unfortunate to define them in `write_buffer` and therefore force unrelated code to depend on things like rdkafka. I therefore lifted and renamed WriteMeta and DbWrite into a new crate called dml

### CODE REVIEW

File: dml/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+use data_types::database_rules::{ShardConfig, ShardId, Sharder};
+use data_types::partition_metadata::{StatValues, Statistics};
+use data_types::sequence::Sequence;
+use mutable_batch::MutableBatch;
+use time::Time;
+use trace::ctx::SpanContext;
+
+/// Metadata information about a DML operation
+#[derive(Debug, Default, Clone)]
+pub struct DmlMeta {
```
Comment (tustvold) [inline dml/src/lib.rs:25]: This was previously called WriteMeta

File: dml/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+/// A DML operation
+#[derive(Debug, Clone)]
+pub enum DmlOperation {
+    /// A write operation
+    Write(DmlWrite),
+}
+
+/// A collection of writes to potentially multiple tables within the same database
+#[derive(Debug, Clone)]
+pub struct DmlWrite {
```
Comment (tustvold) [inline dml/src/lib.rs:95]: This was previously called DbWrite

File: dml/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+    /// Returns the number of bytes read from the wire if relevant
+    pub fn bytes_read(&self) -> Option<usize> {
+        self.bytes_read
+    }
+}
+
+/// A DML operation
+#[derive(Debug, Clone)]
+pub enum DmlOperation {
```
Comment (tustvold) [inline dml/src/lib.rs:88]: This is what the write buffer will "sequence". To be added in a separate PR

Comment (alamb) [reply dml/src/lib.rs:88]: specifically `Delete` will be added in a separate PR 👍

**Review by alamb — APPROVED** [review]
Looks like a nice refactor to me
